ansible/verify_config.py

```diff
@@ -0,0 +1,768 @@
+            return False
+
+        return True
+
+    def _check_if_proxy_is_using(self):
+        group_vars_env = Utility.parse_yml(Config.GROUP_VARS_ENV_FILE)
+
+        env_vars = ["http_proxy", "https_proxy"]
+
+        for var in env_vars:
+            if var not in os.environ:
+                continue
+
+            if var not in group_vars_env or var not in group_vars_env['proxy_env']:
+                self.assertion.add_error_details(
+                    f"'{Formatting.red(var)}' is detected in environment variables "
+                    f"but not in '{Config.GROUP_VARS_ENV_FILE}'",
+                )
+                return False
+
+            if group_vars_env[var] != os.environ[var]:
+                self.assertion.add_error_details(
+                    f"environment '{Formatting.red(var)}' is not the same "
+                    f"as declared in '{Config.GROUP_VARS_ENV_FILE}': "
+                    f"{os.environ[var]} != {group_vars_env[var]}",
+                )
+                return False
+
+        return True
+
+
+class TestBedValidator(Validator):
+
+    def __init__(self, testbed_name):
+        super().__init__()
+        self.testbed_name = testbed_name
+
+    def validate(self):
+        from devutil.devices.factory import init_testbed_sonichosts
+
+        if not self.testbed_name:
+            return True
+
+        yml_config = next(
+            filter(lambda tb: tb['conf-name'] == self.testbed_name, Utility.parse_yml(Config.TESTBED_FILE))
+        )
+
+        sonic_hosts = init_testbed_sonichosts(yml_config['inv_name'], self.testbed_name)
+
+        self.assertion.assert_true(
+            lambda: self._check_if_duts_are_reachable(sonic_hosts),
+            reason=f"Devices are not reachable for testbed '{Formatting.red(self.testbed_name)}'. "
+                   "Please check your proxy configs",
+        )
+
+    def _check_if_duts_are_reachable(self, sonic_hosts):
+        _, result = sonic_hosts.reachable()
+        is_valid = True
+
+        for dut_result in result.values():
+            if not dut_result["reachable"] or dut_result["failed"]:
+                is_valid = False
+                self.assertion.add_error_details(
+                    f"The following device is unreachable '{Formatting.red(dut_result['hostname'])}'. "
+                    f"Message: '{Formatting.yellow(dut_result['msg'])}'",
+                )
+
+        return is_valid
+
+
+def main(args):
+    if args.testbed_file:
+        Config.TESTBED_FILE = args.testbed_file
+
+    if args.vm_file:
+        Config.VM_FILE = args.vm_file
+
+    validators = []
+
+    if args.target:
+        validators.extend([TestBedValidator(args.target)])
+    else:
+        validators.extend([
+            DockerRegistryValidator(),
+            TestbedValidator(),
+            InventoryNameValidator(),
+            FanoutLinkValidator(),
+            HostNetworkValidation()
+        ])
```

Comment:
With this `if...else...` logic here, does it mean we will only check the connectivity as long as there is `-tb` in the argument?

Reply:
We need to provide `-tb testbed-name` not just `-tb` for it to work. When specifying a testbed, i'm only checking for connection now using ansible. I was thinking to also check using `TestbedValidator()` as well but that validator will need to go through all the testbed to find a conflict and it will still logs out all the validation error from other testbed. 

For inventory and others, if we're not specifying it correctly, ansible will not work as well so it's already included in the connection checking

Reply:
Ok then let's modify/remove the comment at the beginning of this file
```
To verify a single testbed’s connectivity using specific testbed and VM files:
- python3 verify_config.py -t <testbed-file> -m <vm-file> -tb <testbed-name>
```